dali/pipeline/data/tensor_vector.cc

```diff
@@ -766,22 +755,79 @@ TensorVector<Backend> &TensorVector<Backend>::operator=(TensorVector<Backend> &&
     tl_ = std::move(other.tl_);
     type_ = other.type_;
     sample_dim_ = other.sample_dim_;
-    views_count_ = other.views_count_.load();
     tensors_ = std::move(other.tensors_);
-    for (auto &t : tensors_) {
-      if (t) {
-        if (auto *del = std::get_deleter<ViewRefDeleter>(t->data_)) del->ref = &views_count_;
-      }
-    }
 
-    other.views_count_ = 0;
     other.curr_num_tensors_ = 0;
     other.tensors_.clear();
   }
   return *this;
 }
 
 
+// This is to check if we are actually laid down in contiguous memory
+// TODO(klecki): make this internal and name it something like: IsContiguouslyStored?
+template <typename Backend>
+bool TensorVector<Backend>::IsContiguousTensor() const {
+  if (num_samples() == 0 || shape().num_elements() == 0) {
+    return true;
+  }
+  // If we are using contiguous representation in this case we can safely return
+  if (IsContiguous()) {
+    return true;
+  }
+  const uint8_t *base_ptr = static_cast<const uint8_t *>(tensors_[0]->raw_data());
+  size_t size = type_info().size();
+
+  for (int i = 0; i < num_samples(); ++i) {
+    if (base_ptr != tensors_[i]->raw_data()) {
+      return false;
+    }
+    base_ptr += tensors_[i]->shape().num_elements() * size;
+  }
+  return true;
+}
+
+
+template <typename Backend>
+bool TensorVector<Backend>::IsDenseTensor() const {
+  return IsContiguous() && is_uniform(shape());
+}
+
+
+template <typename Backend>
+Tensor<Backend> TensorVector<Backend>::AsReshapedTensor(const TensorShape<> &new_shape) {
+  DALI_ENFORCE(num_samples() > 0,
+               "To create a view Tensor, the batch must have at least 1 element.");
+  DALI_ENFORCE(IsValidType(type()),
+               "To create a view Tensor, the batch must have a valid data type.");
+  DALI_ENFORCE(
+      shape().num_elements() == new_shape.num_elements(),
+      make_string("To create a view Tensor, requested shape need to have the same volume as the "
+                  "batch, requested: ",
+                  new_shape.num_elements(), " expected: ", shape().num_elements()));
+  Tensor<Backend> result;
+  result.ShareData(unsafe_sample_owner(*tl_, 0), tl_->capacity(),
+                   tl_->is_pinned(), new_shape, type(), device_id(), order());
+  auto result_layout = GetLayout();
+  if (!GetLayout().empty()) {
+    result_layout = TensorLayout("N") + result_layout;
+  }
+  result.SetLayout(result_layout);
```

Comment:
If we reshape, we can only use the source layout if the output shape has one extra (sample) dim.
```suggestion
  auto result_layout = GetLayout();
  if (result_layout.ndim() + 1 == shape.sample_dim()) {
    result_layout = TensorLayout("N") + result_layout;
    result.SetLayout(result_layout);  
  }
```
We could possibly copy the sample layout to the destination when the target shape has the same number of dimensions as our sample, but that is guessing.

Reply:
Done. Good spot.